Makefile

```diff
@@ -1,36 +1,177 @@
-SOURCE_COMMIT := $(shell git log -1 --pretty=format:"%H")
-SOURCE_BRANCH := $(shell git branch | grep \* | cut -d ' ' -f2)
-TAG := target/flottbot
+PWD := $(shell pwd)
+GOPATH := $(shell go env GOPATH)
+GIT_HASH := $(shell git log -1 --pretty=format:"%H")
+VERSION := $(shell git describe --tags --always)
+GOARCH := $(shell go env GOARCH)
+GOOS := $(shell go env GOOS)
+BUILD_LDFLAGS := -s -w
+BUILD_LDFLAGS += -X github.com/target/flottbot/version.Version=${VERSION}
+BUILD_LDFLAGS += -X github.com/target/flottbot/version.GitHash=${GIT_HASH}
+GOLANGCI_LINT_VERSION := "v1.23.8"
 
-.PHONY: all docker
+DOCKER_IMAGE ?= "target/flottbot"
+DOCKER_FLAVORS ?= golang ruby python
+PLATFORMS ?= linux/amd64 darwin/amd64 windows/amd64
 
-build:
-	go build -v -a -o flottbot cmd/flottbot/main.go
+# some helpers for building for each platform
+p = $(subst /, ,$@)
+os = $(word 1, $(p))
+arch = $(word 2, $(p))
 
-run: build
-	./flottbot
+.PHONY: all
+all: test build
+
+# ┌┬┐┌─┐┌─┐┌┬┐
+#  │ ├┤ └─┐ │ 
+#  ┴ └─┘└─┘ ┴ 
+.PHONY: validate
+validate: getdeps fmt vet lint
+
+.PHONY: getdeps
+getdeps:
+	@mkdir -p ${GOPATH}/bin
+	@which golangci-lint 1>/dev/null || \
+		(echo "Installing golangci-lint" && \
+		curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | \
+		sh -s -- -b $(go env GOPATH)/bin $(GOLANGCI_LINT_VERSION))
+
+.PHONY: lint
+lint:
+	@echo "Running $@ check"
+	@GO111MODULE=on golangci-lint run
+
+.PHONY: fmt
+fmt:
+	@echo "Running $@ check"
+	@GO111MODULE=on go fmt ./...
+
+.PHONY: vet
+vet:
+	@echo "Running $@ check"
+	@GO111MODULE=on go vet ./...
+
+.PHONY: tidy
+tidy:
+	@echo "Running $@"
+	@go mod tidy
+
+.PHONY: ensure-go-acc
+ensure-go-acc:
+	@which go-acc 1>/dev/null || \
+		(echo "Installing go-acc" && \
+		GO111MODULE=off go get -u github.com/ory/go-acc)
+
+.PHONY: test
+test:
+	@echo "Running unit tests"
+	@go test ./...
+
+.PHONY: test-race
+test-race: ensure-go-acc
+	@echo "Running unit tests with -race"
+	@go-acc ./... --ignore config-example -o coverage.out
+
+.PHONY: clean
+clean: validate tidy
+	@echo "Running $@ tasks"
+	-rm -v ./flottbot*
+	-rm -v ./debug
+
+# ┌┐ ┬ ┬┬┬  ┌┬┐
+# ├┴┐│ │││   ││
+# └─┘└─┘┴┴─┘─┴┘
+.PHONY: build
+build: clean
+	@echo "Building flottbot binary to './flottbot'"
+	@GO111MODULE=on go build -a \
+		-ldflags '$(BUILD_LDFLAGS)' -o $(PWD)/flottbot ./cmd/flottbot
 
-docker:
-	docker build \
-		--build-arg "SOURCE_BRANCH=$(SOURCE_BRANCH)" \
-		--build-arg "SOURCE_COMMIT=$(SOURCE_COMMIT)" \
+.PHONY: build-cross
```

Comment:
Could consider switching to [goreleaser](https://github.com/goreleaser/goreleaser-action) to simplify the building and releasing workflow.

Reply:
i did consider it.. was a little hesitant to throw another config file in the root for goreleaser.. i'll take a look though as i'm not super familiar with it

Reply:
alright, now with goreleaser :D